Pancan new activities

YuriyPavlovBI opened the pull request:
There are 3 new activities added to the cancan study and some platform-level functionality needed for these activities to work:
https://broadinstitute.atlassian.net/browse/DDP-6957
https://broadinstitute.atlassian.net/browse/DDP-6958
https://broadinstitute.atlassian.net/browse/DDP-6959

- [x] https://github.com/broadinstitute/ddp-study-server/pull/1050 Added reevaluation of the block visibilities for the parent activity when child activity is being added/removed. (In this case, we need to hide "I don't have children" checkbox if at least one child is added) 

- [x] https://github.com/broadinstitute/ddp-study-server/pull/1051 New question type added: Activity Instance Select - allows user to select an instance of some configured list of activities for that particular question. In this case it is used to select which parent side some relative is on. Stored value is the activity instance guid, shown value is a rendered activity instance name. Options can be obtained using `/user/{user}/studies/{study}/activity-instance-select/{stableId}` endpoint.

- [x] https://github.com/broadinstitute/ddp-study-server/pull/1052 Support for pre-selected answers. New attribute is introduced to the pick list options: **_default_**. If question has at least one default option, the default answer will be automatically written right after activity instance creation. 

- [x] https://github.com/broadinstitute/ddp-study-server/pull/1059 Implement new matrix question type

- [x] https://github.com/broadinstitute/ddp-study-server/pull/1064 DDP-7085 Extend API to support disabling activity block

YuriyPavlovBI committed:
DDP-6958 Add Diet and lifestyle survey

YuriyPavlovBI committed:
DDP-6958 Add Diet and lifestyle survey

YuriyPavlovBI committed:
DDP-6959 Add: Stool kit survey

YuriyPavlovBI committed:
DDP-6959 Add: Stool kit survey

YuriyPavlovBI committed:
DDP-6958 Add Diet and lifestyle survey

YuriyPavlovBI committed:
DDP-6958 Add Diet and lifestyle survey

YuriyPavlovBI committed:
Added family history

YuriyPavlovBI committed:
Added family history

YuriyPavlovBI committed:
Merge branch 'pancan_new_activities' of github.com:broadinstitute/ddp-study-server into pancan_new_activities

# Conflicts:
#	study-builder/studies/pancan/i18n/es.conf

YuriyPavlovBI committed:
Family History Adjustments

YuriyPavlovBI committed:
Family History i18n

YuriyPavlovBI committed:
Family History i18n

YuriyPavlovBI committed:
Family History i18n

YuriyPavlovBI committed:
Family History i18n

YuriyPavlovBI committed:
Merge branch 'develop' into pancan_new_activities

YuriyPavlovBI committed:
Family History i18n

YuriyPavlovBI committed:
Family History Adjustments

YuriyPavlovBI committed:
DDP-6957 Add Family History Survey

New question type added to allow user to select from a list of previously given answers (e.g. parent "nickname")

# Conflicts:
#	study-builder/studies/pancan/family-history/snippets/family-history-block-question-side.conf

YuriyPavlovBI committed:
DDP-6957 Add Family History Survey
New question type added to allow user to select from a list of previously given answers (e.g. parent "nickname")

YuriyPavlovBI committed:
Merge branch 'develop' into pancan_new_activities

YuriyPavlovBI committed:
DDP-6957 Add Family History Survey
Some fixes

YuriyPavlovBI committed:
DDP-6957 Add Family History Survey
Some fixes

YuriyPavlovBI committed:
DDP-6957 Add Family History Survey
Some fixes

YuriyPavlovBI committed:
DDP-6957 Add Family History Survey

Changing the concept: DYNAMIC question type becomes ACTIVITY_INSTANCE_SELECT question

YuriyPavlovBI committed:
DDP-6957 Add Family History Survey

Changing the concept: DYNAMIC question type becomes ACTIVITY_INSTANCE_SELECT question

# Conflicts:
#	study-builder/studies/pancan/family-history/snippets/family-history-block-question-side-nested.conf

YuriyPavlovBI committed:
Add blockVisibility for parentActivity blocks during creation and deletion of the child instances

YuriyPavlovBI committed:
Add blockVisibility for parentActivity blocks during creation and deletion of the child instances

YuriyPavlovBI committed:
GetActivityInstanceListForActivityInstanceSelectQuestionRouteStandaloneTest fixed

YuriyPavlovBI committed:
activityCodes

ssettipalli commented:
@ YuriyPavlovBI
I see that PR already has 141 files . I see that more changes will be included in this PR. Please try to create separate PR's for different tickets. That way we can keep the PR's to reasonable and reviewable size.

I see that you added a new Question Type : ActivityInstanceSelect 
We usually have a separate ticket and PR for core platform changes like new QuestionType and study related changes.
cc @MocanaAtBroad

MocanaAtBroad commented:
Hi Yuriy!
Looks like you have been extremely busy! There is a ton of stuff here and it is going to be very difficult to review it all.

I am going to request that you break this P.R. in chunks, starting with just the proposed API changes, including explanations on why the API changes are needed/how they address new requirements/etc.

After we have processed that P.R. we can take a look at the back-end and front-end code changes needed to implement the API changes. If you are adding a new question type it would be very useful to create an activity in basil study that shows the new question type in action. Doesn't have to be anything elaborate: just an activity that has one or more variations of the question type in action.

Finally, in separate PRs, let's consider the study configuration changes. If you could do separate PRs for each survey, that would be best as it limits the size of the chunk that needs to be inspected.

Reach out via Slack if you have questions. Thanks!

YuriyPavlovBI committed:
Fixed Parent2 activity name

YuriyPavlovBI commented:
Sounds reasonable! I've created 2 PRs, one per feature. Let's keep it PR as a "parent" PR. Once "child" PRs worked out and merged this one will contain just a configuration changes and we will split them up by activities if needed.

YuriyPavlovBI committed:
Fixed activity instance select snippet

YuriyPavlovBI committed:
Fixed activity instance select snippet

YuriyPavlovBI committed:
TODOs refined

YuriyPavlovBI committed:
DDP-7074 Support for pre-selected answers

# Conflicts:
#	pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java
#	pepper-apis/src/main/resources/changelog-master.xml

YuriyPavlovBI committed:
DDP-7074 Support for pre-selected answers

YuriyPavlovBI committed:
DDP-7074 Support for pre-selected answers
Configuration

YuriyPavlovBI committed:
Test fixed

YuriyPavlovBI committed:
Test fixed

YuriyPavlovBI committed:
Tests fixed

YuriyPavlovBI committed:
Tests fixed

YuriyPavlovBI committed:
Tests fixed

YuriyPavlovBI committed:
Tests fixed

YuriyPavlovBI committed:
Tests fixed

YuriyPavlovBI committed:
Tests fixed

YuriyPavlovBI committed:
Review fixes

YuriyPavlovBI committed:
Review fixes

YuriyPavlovBI committed:
Merge branch 'develop' into activity-instance-select-question-type

YuriyPavlovBI committed:
Merge branch 'develop' into DDP-7074-default-picklist-options

YuriyPavlovBI committed:
Conflict merge fix

YuriyPavlovBI committed:
isDefault field was hidden

YuriyPavlovBI committed:
Merge branch 'develop' into DDP-7074-default-picklist-options

YuriyPavlovBI committed:
Merge branch 'develop' into DDP-7074-default-picklist-options

YuriyPavlovBI committed:
Merge branch 'develop' into activity-instance-select-question-type

YuriyPavlovBI committed:
Review fix

YuriyPavlovBI committed:
Review fix

YuriyPavlovBI committed:
Merge branch 'develop' into activity-instance-select-question-type

YuriyPavlovBI committed:
Update pepper-apis/docs/specification/src/components/schemas.yml

Co-authored-by: Blake (he/him) <bskinner@broadinstitute.org>

YuriyPavlovBI committed:
Merge branch 'parent-activity-block-visibility-reevaluation' into pancan_new_activities

# Conflicts:
#	pepper-apis/docs/specification/src/components/responses.yml
#	pepper-apis/docs/specification/src/components/schemas.yml
#	pepper-apis/docs/specification/src/endpoints/user.studies.activities.yml
#	pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateActivityInstanceRoute.java
#	pepper-apis/src/main/java/org/broadinstitute/ddp/route/DeleteActivityInstanceRoute.java
#	pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java

YuriyPavlovBI committed:
Merge branch 'DDP-7074-default-picklist-options' into pancan_new_activities

# Conflicts:
#	pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistQuestionDao.java
#	pepper-apis/src/main/java/org/broadinstitute/ddp/util/QuestionUtil.java
#	pepper-apis/src/main/resources/changelog-master.xml

YuriyPavlovBI committed:
Cleanup imports

YuriyPavlovBI committed:
Merge failures cleanup

YuriyPavlovBI committed:
Merge branch 'develop' into pancan_new_activities

YuriyPavlovBI committed:
Fixed substitutions for Family History

YuriyPavlovBI committed:
Review fixes

YuriyPavlovBI committed:
Fixed docs

YuriyPavlovBI committed:
Merge branch 'develop' into activity-instance-select-question-type

YuriyPavlovBI committed:
Activity Instance Select question example

YuriyPavlovBI committed:
Merge branch 'activity-instance-select-question-type' into pancan_new_activities

# Conflicts:
#	pepper-apis/docs/specification/src/components/responses.yml
#	pepper-apis/docs/specification/src/components/schemas.yml
#	pepper-apis/src/main/java/org/broadinstitute/ddp/export/collectors/ActivityInstanceSelectQuestionFormatStrategy.java
#	pepper-apis/src/main/java/org/broadinstitute/ddp/export/collectors/ActivityResponseCollector.java
#	pepper-apis/src/main/java/org/broadinstitute/ddp/route/GetActivityInstanceListForActivityInstanceSelectQuestionRoute.java
#	pepper-apis/src/main/java/org/broadinstitute/ddp/route/PatchFormAnswersRoute.java
#	pepper-apis/src/main/resources/changelog-master.xml
#	pepper-apis/src/test/java/org/broadinstitute/ddp/export/collectors/ActivityInstanceSelectQuestionFormatStrategyTest.java

YuriyPavlovBI committed:
DDP-7085 Extend API to support disabling activity block

YuriyPavlovBI committed:
DDP-7085 Extend API to support disabling activity block
Parent activities non-deletable

YuriyPavlovBI committed:
Fixed typo

bskinner commented on study-builder/studies/pancan/family-history/family-history-top-level.conf:
```diff
@@ -0,0 +1,484 @@
+            "hideNumber": true,
+            "groups": [],
+            "validations": [],
+            "promptTemplate": {
+              "templateType": "HTML",
+              "templateText": "",
+              "variables": []
+            },
```
What's going on here? It looks like this question is setup to have an empty prompt.

YuriyPavlovBI replied:
That's correct! This block is just a single checkbox. The "I do not have any information..." checkbox here:
<img width="1195" alt="image" src="https://user-images.githubusercontent.com/71830706/142260133-cb30d34f-2fe7-4781-980c-5ba3f3e09197.png">

bskinner replied:
In that case, the structure needs to be reworked. If the "I do not have any information..." text is intended to be presented with the checkbox, it should be included as part of the question's data. At the moment, there is no way to know these two elements should be presented together, and only happens to work because of the way the Angular client renders things.

bskinner commented on study-builder/studies/pancan/family-history/family-history-additional-details.conf:
```diff
@@ -0,0 +1,84 @@
+            "stableId": "OTHER_FACTORS_CANCER_RISK",
+            "inputType": "ESSAY",
+            "hideNumber": true,
+            "promptTemplate": {
+              "templateType": "TEXT",
+              "templateText": "",
+              "variables": []
+            },
```
This seems like a similar situation to my [other comment](https://github.com/broadinstitute/ddp-study-server/pull/1049#discussion_r751290694). There is a question, which is not part of a `COMPOSITE`, which lacks a prompt.

YuriyPavlovBI replied:
This is also made on purpose. Previous blocks contains it all.
This actually is the following section:
<img width="973" alt="image" src="https://user-images.githubusercontent.com/71830706/142260781-2574a6af-9534-4e4c-9a66-4c85679fbf9d.png">

bskinner replied:
Same thing as my other comment, this design needs to be reworked. A `CONTENT` block may not be used as a replacement for presenting a question's prompt.

YuriyPavlovBI replied:
Moved text to the prompt

YuriyPavlovBI committed:
Refactored grids for new question type usage

YuriyPavlovBI committed:
Removed _GRID suffix

YuriyPavlovBI committed:
Spanish translations

YuriyPavlovBI committed:
Merge branch 'develop' into pancan_new_activities

YuriyPavlovBI committed:
Structure changes

YuriyPavlovBI committed:
Merge branch 'develop' into pancan_new_activities

YuriyPavlovBI committed:
Merge branch 'develop' into pancan_new_activities

YuriyPavlovBI committed:
Blood consent, update script

YuriyPavlovBI committed:
Review fixes

aleksey-kuznetsov reviewed (dismissed):
Let's go!

YuriyPavlovBI committed:
Update UserAuthCheckFilter.java

YuriyPavlovBI committed:
Update CreateActivityInstanceRoute.java

YuriyPavlovBI committed:
Update CreateActivityInstanceRoute.java

YuriyPavlovBI committed:
Update AuthPathRegexUtil.java
